builtin/aws/lambda/platform.go

```diff
@@ -0,0 +1,706 @@
+package lambda
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/subtle"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/crypto/ssh"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/lambda"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint/builtin/aws/ecr"
+	"github.com/hashicorp/waypoint/builtin/aws/utils"
+	"github.com/pkg/errors"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/docs"
+	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
+)
+
+// Platform is the Platform implementation for AWS Lambda
+type Platform struct {
+	config Config
+}
+
+// ConfigSet is called after a configuration has been decoded
+// we can use this to validate the config
+func (p *Platform) ConfigSet(config interface{}) error {
+	_, ok := config.(*Config)
+	if !ok {
+		// this should never happen
+		return fmt.Errorf("Invalid configuration, expected *lambda.Config, got %s", reflect.TypeOf(config))
+	}
+
+	return nil
+}
+
+// Config implements Configurable
+func (p *Platform) Config() (interface{}, error) {
+	return &p.config, nil
+}
+
+// DeployFunc implements component.Platform
+func (p *Platform) DeployFunc() interface{} {
+	return p.Deploy
+}
+
+// DestroyFunc implements component.Destroyer
+func (p *Platform) DestroyFunc() interface{} {
+	return p.Destroy
+}
+
+// ExecFunc implements component.Execer
+func (p *Platform) ExecFunc() interface{} {
+	return p.Exec
+}
+
+// ValidateAuthFunc implements component.Authenticator
+func (p *Platform) ValidateAuthFunc() interface{} {
+	return p.ValidateAuth
+}
+
+// AuthFunc implements component.Authenticator
+func (p *Platform) AuthFunc() interface{} {
+	return p.Auth
+}
+
+func (p *Platform) Auth() error {
+	return nil
+}
+
+// TODO sort out the right way to validate AWS config and use it in all the AWS types
+func (p *Platform) ValidateAuth(
+	ctx context.Context,
+	log hclog.Logger,
+	src *component.Source,
+	ui terminal.UI,
+) error {
+	return nil
+}
+
+const (
+	// The default amount of memory to give to the function invocation, 256MB
+	DefaultMemory = 256
+
+	// How long the function should run before terminating it.
+	DefaultTimeout = 60
+)
+
+const lambdaRolePolicy = `{
+	"Version": "2012-10-17",
+	"Statement": [
+		{
+			"Effect": "Allow",
+			"Principal": {
+				"Service": "lambda.amazonaws.com"
+			},
+			"Action": "sts:AssumeRole"
+		}
+	]
+}`
+
+// SetupRole creates an IAM role that will be used by the Lambda function. The role
+// has the basic lambda execution rules attached to it.
+func (p *Platform) SetupRole(L hclog.Logger, app *component.Source, sess *session.Session) (string, error) {
```

Comment:
feels like this should be unexported to better convey this is an internal helper